[SourceKit] Include operation kind in compile notifications

DavidGoldman opened the pull request:
- Compile will start / did finish notifications now include the operation kind (perform sema or code complete).

- Motivation: see [this swift forums post](https://forums.swift.org/t/sourcekit-lsp-file-status-ux/35947)

DavidGoldman committed:
[SourceKit] Include operation kind in compile notifications

- Compile will start / did finish now includes the operation
  kind (perform sema or code complete).
- See also
  https://forums.swift.org/t/sourcekit-lsp-file-status-ux/35947

DavidGoldman commented:
@swift-ci please test

benlangmuir commented on utils/gyb_sourcekit_support/UIDs.py:
```diff
@@ -182,6 +182,7 @@ def __init__(self, internal_name, external_name):
     KEY('CompletionCheckDependencyInterval',
         'key.completion_check_dependency_interval'),
     KEY('AnnotatedTypename', 'key.annotated.typename'),
+    KEY('CompileOperation', 'key.compileoperation'),
```
We're a bit inconsistent about this, but I would suggest `key.compile_operation`

DavidGoldman replied:
Done

benlangmuir commented on tools/SourceKit/tools/sourcekitd/lib/API/Requests.cpp:
```diff
@@ -3181,6 +3181,22 @@ class CompileTrackingConsumer final : public trace::TraceConsumer {
 } // end anonymous namespace
 
+static UIdent getUIDForOperationKind(trace::OperationKind OpKind) {
+  static UIdent CompileOperationPerformSema("source.compile.operation.perform-sema");
```
Maybe we should just leave off the kind when it's from the ASTManager? This name is not specific (every AST build except code-completion uses performSema), and it is meaningless to  a user.

DavidGoldman replied:
Done

DavidGoldman committed:
Fix up compile operation kind

- `key.compile_operation` instead of `key.compileoperation`
- Make the operation kind optional (nothing emitted for perform sema)

DavidGoldman commented:
@swift-ci please test

DavidGoldman commented:
@swift-ci please test

owenv commented on utils/gyb_sourcekit_support/UIDs.py:
```diff
@@ -182,6 +182,7 @@ def __init__(self, internal_name, external_name):
     KEY('CompletionCheckDependencyInterval',
         'key.completion_check_dependency_interval'),
     KEY('AnnotatedTypename', 'key.annotated.typename'),
+    KEY('CompileOperation', 'key.compile_operation'),
```
It doesn't need to be part of this PR, but could you document the new key in tools/SourceKit/docs/Protocol.md?

benlangmuir commented on tools/SourceKit/tools/sourcekitd/lib/API/Requests.cpp:
```diff
@@ -3181,6 +3181,21 @@ class CompileTrackingConsumer final : public trace::TraceConsumer {
+      return CompileOperationIndexSource;
+    case trace::OperationKind::CodeCompletion:
+      return CompileOperationCodeCompletion;
+    case trace::OperationKind::All:
```
You can do something like 
```c++
default:
  llvm_unreachable("unknown operation kind");
```

instead of handling "All", which is not a valid case, it's only used when treating the operation as a bitset for desiredOperations.

DavidGoldman replied:
Done

benlangmuir reviewed (commented):
One minor thing, but otherwise looks good to me.

DavidGoldman committed:
[SourceKit] Don't handle OperationKind::all

DavidGoldman commented:
@swift-ci please test

DavidGoldman commented:
@swift-ci please test

benlangmuir commented:
@swift-ci please test Windows
